csharp/EmsApi.Client/V2/Wrappers/EmsApiRouteWrapper.cs

```diff
@@ -0,0 +1,135 @@
+﻿
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace EmsApi.Client.V2.Wrappers
+{
+	/// <summary>
+	/// An API route wrapper is intended to convert the raw IEmsApi interface calls
+	/// to something that is a little more natural to work with inside .NET applications.
+	/// </summary>
+	public abstract class EmsApiRouteWrapper
+	{
+		/// <summary>
+		/// Creates a new instance of a route wrapper.
+		/// </summary>
+		/// <param name="api">
+		/// The raw API interface to make calls to.
+		/// </param>
+		public EmsApiRouteWrapper( IEmsApi api )
+		{
+			m_api = api;
+		}
+
+		/// <summary>
+		/// The reference to the raw api interface.
+		/// </summary>
+		/// <remarks>
+		/// This is private so classes which derive from this one must use the
+		/// <seealso cref="ContinueWithExceptionSafety{TRet}(Func{IEmsApi, Task{TRet}})"/>
+		/// method instead.
+		/// </remarks>
+		private IEmsApi m_api;
+
+		/// <summary>
+		/// Event that is fired whenever a low level API exception occurs.
+		/// </summary>
+		/// <remarks>
+		/// This is static so that this same event will fire for all derived wrapper classes.
+		/// </remarks>
+		internal static event EventHandler<ApiExceptionEventArgs> ApiMethodFailedEvent;
+
+		/// <summary>
+		/// Executes a delegate that returns a task which accesses the API interface.
+		/// The task forwards all exceptions through the <seealso cref="ApiMethodFailedEvent"/>.
+		/// </summary>
+		/// <param name="apiFunc">
+		/// The delegate for a function to execute.
+		/// </param>
+		protected Task<TRet> ContinueWithExceptionSafety<TRet>( Func<IEmsApi, Task<TRet>> apiFunc )
+		{
+			// All the Refit methods return tasks, so we need to handle exceptions
+			// as continuations.
+			return apiFunc( m_api );
+
+			/*
+			var api = apiFunc( m_api );
+			var safeTask = api.ContinueWith( HandleApiException, TaskContinuationOptions.OnlyOnFaulted );
+			return safeTask;
+			*/
+		}
+
+		/// <summary>
+		/// Checks the input enumerable for null, and returns an empty enumerable instead
+		/// if it is null.
+		/// </summary>
+		/// <remarks>
+		/// This is intended to help handle cases where an async call might return null
+		/// when it fails.
+		/// </remarks>
+		/// <typeparam name="T">
+		/// The type of value that the enumerable returns.
+		/// </typeparam>
+		/// <param name="input">
+		/// The enumerable to check.
+		/// </param>
+		protected IEnumerable<T> EnsureEnumerableNotNull<T>( IEnumerable<T> input )
+		{
+			return input == null ? Enumerable.Empty<T>() : input;
+		}
+
```

Comment:
Naming bikeshedding: this doesn't ensure the enumerable is not null, it returns a non-null enumerable. The name feels awkward to me - maybe SafeAccessEnumerable? not sure.
I'd also make this a static function to communicate that it doesn't access class values.